### ISSUE TITLE
Consider queue per scribe pattern

### ISSUE THREAD

**Issue by MichaelXavier** [Soostone/katip#49]
Discussed in some detail in #46 but it was out of scope for the topic. The idea is that using a chan or queue of some sort is seemingly the right choice for most scribes. Rather than having scribes have to implement their own half-baked queueing strategy, katip should do it for them. We probably don't even have to change the type of scribe since its just an IO action that receives a log item. We can just wire that action up to a queue checking loop.

We however may want to break API to allow the user to set per-scribe buffering limits if there isn't a nice way of defaulting that. We'd also want to consider a way to encode finalization behavior. For example, I think katip-elasticsearch may still keep its own internal queue since it has the concept of multiple worker threads writing to elasticsearch. Perhaps we should also encode the finalizer into the scribe type itself so that there is only 1 finalizer, the katip finalizer that will drian *its* queues then call the blocking finalizer to drain each scribe before returning.

Tagging in @adinapoli

**Comment by adinapoli** [comment]
> Perhaps we should also encode the finalizer into the scribe type itself so that there is only 1 finalizer, the katip finalizer that will drian its queues then call the blocking finalizer to drain each scribe before returning.

I like this! It's pretty much was I was proposing along the lines in #48, and I'm glad we had the same idea 😉

**Comment by MichaelXavier** [comment]
@adinapoli @ozataman I've got a rough sketch of this [here](https://github.com/Soostone/katip/compare/49-queue-subscriber?expand=1). I still need to do some cleanup, testing, and documentation but here's the gist:

1. Added a finalizer field to scribe. I tried it without this but it got sort of messy. Its a useful signal to library implementers that the scribe record comes with a finalizer field so that they have to actively opt out if they don't need one.
2. Added a settings record for each scribe. Right now it just takes a queue bound and defaults to 4096. I plan on not exporting the constructor but exporting lenses/fields like we do elsewhere to avoid breaking changes on additive fields later.
3. Made all of the LogEnv modifying functions like `clearScribes` and `removeScribe` run in IO since they need to be concerned about finalizers now. `clearScribe` is effectively katip's shutdown signal. It may make sense before release to rename this to something that indicates that better and to add examples showing how this is done.
4. Added a bounded unagi-chan channel per scribe. Each scribe added fires off a thread that loops on messages. Writes when the queue is full get dropped on the floor.
5. I took @adinapoli's pattern and on katip shutdown, a `PoisonPill` message gets send to each worker in a blocking fashion. This is because unagi-chan lacks a structure that supports closing. PoisonPill signifies it will be the last message to be processed and the worker will quit. After each katip worker quits, it runs the scribe's blocking finalizer so that it can clean up and finish flushing messages. Thus we get a nice cascading shutdown where all writes are flushed.
6. I rolled back @adinapoli's buffered Handle scribe. Because katip does the buffering internally all scribes can afford to be fairly simple and not necessarily worry about buffering. The elasticsearch worker still uses a queue because internally it supports concurrent writes to elasticsearch (which makes more sense than concurrent writes to a handle). I will probably end up using an unagi-chan bounded channel for that since it doesn't make sense to use 2 queueing libraries in the same repo. I will probably also drop the queue size field from the scribe config since its specified higher up and it doesn't make sense to specify that twice. It does seem kinda funky to have a katip queue funnel right into another queue for workers. I wonder if it would be appropriate to lean on that queueing and just have the interface for elasticsearch be an mvar that its workers fight over. The katip queue for the ES scribe would blockingly write into that mvar. Worker threads would take the latest write out of that mvar and the queue above would be able to flush. This way we'd end up using worst case `message size * queue size + message size` memory vs with two queues, up to `message size * queue size * 2` memory. I'm just not sure if contention on an mvar has worse constant factors than on a chan.

**Comment by adinapoli** [comment]
Hey @MichaelXavier, thank you for this! I gave a quick skim last night and it looks very good, apart maybe for a brittle story for catching exceptions (which I am admittedly a noob on the topic myself). The new-in-town [safe-exceptions](https://hackage.haskell.org/package/safe-exceptions) library seems to be these days one of the best, so maybe we should give it a go? I'm going to put some comments on the PR as well, but nothing substantial 😉 

Something which I will add, which can be totally obvious to you guys, is that we should really aim to give the user the "last word" about running the finalisers (and that's why I am so pleased to see Michael's `clearScribes` abstraction). **In other terms, let's not make the mistake of having the library running those for them.**

This is better explained with a tale from the trenches: after my #48 patch, I started using the improved `Scribe` in a project, and I have a monad for which I have defined `Katip` and `KatipContext` on, as well as a small record for the bookeeping, something like this:

```
data MyEnv = MyEnv {
    logEnv :: LogEnv
  , ctx       :: LogContexts
  , finalisers :: [IO ()]
}
``` 

Now, the reason why that `finalisers` field is there is because I have 2 kind of programs using this monad: one is short lived, and the monad is threaded throughout the program. Differently put, there is only one `runXXX`, which reifies it back into `IO ()`. The other is a server which can invoke multiple times `runXXX`. Cutting all the details, I was running the finalisers (**including the Handle Scribe ones**) as a post-hook inside `runXXX`, which was correct for _some_ scenario which I won't cover for brevity, but for the `Handle` Scribe was a totally-lame design choice from me, as due to the fact `MyEnv` is shared across multiple threads in my server, and there are multiple calls to `runXXX`, soon after the first monad completed back into `IO ()`, my Handle scribes were shutdown and I couldn't basically log anymore 🙈 

Hope this doesn't sound contrived (probably is, sorry!), but that's just a small story to say that if Katip would have imposed WHEN to run the finalisers, my life would have been miserable. Instead, I ended up having a set of finalisers (to be run frequently after each monad completes) and a set of "shutdown actions" which I can invoke only once, and best of all I am in control of the process.

**TL;DR: I think Michael's patch & intentions are exactly right, which are giving users power & control on when run the finalisers, but I was simply reporting a real world scenario to further back-up this choice!**

**Comment by MichaelXavier** [comment]
Thanks for the comments. I think the way Katip was designed makes it fairly difficult for katip to understand its own lifecycle: you add a `Katip` instance to your application's monad that provides a `LogEnv`. This says nothing about when you initialize the `LogEnv` (usually on application start), so it doesn't provide us much of a sensible way to end katip's lifecycle. That's also kind of why I want to rename `clearScribes` because it doesn't communicate to me that its a shutdown sort of action. You'll notice that the reference function `katipNoLoggingT` can't use `clearScribes` anymore because its meant to be a temporary pausing of logging and thus it would be inappropriate to shut down loggers. Thankfully, the fact that `clearScribes` happens in `IO` now would make it impractical for users implementing a no-logging function in their own monad to accidentally use `clearScribes` and permanently cancel logging when they meant it to be temporary.

**Comment by adinapoli** [comment]
> That's also kind of why I want to rename clearScribes because it doesn't communicate to me that its a shutdown sort of action. 

Yea, I see what you mean! I agree that something more "extreme" like `shutdownScribes` or something like that could more effectively communicate the intentions behind that function 😀 

> You'll notice that the reference function katipNoLoggingT can't use clearScribes anymore because its meant to be a temporary pausing of logging and thus it would be inappropriate to shut down loggers.

If we are really desperate about pausing loggers and other things like that we could always (I guess) add things like `pauseScribes`, `resumeScribes` and extend the `WorkerCmd` to have things like `Pause` and `Resume`, but one might argue that could complicate the API and the ease-of-use of the library, and he would probably be right 😅

**Comment by MichaelXavier** [comment]
Resurrecting this issue as I got some time to put more work into this today. In the last few commits I was pretty detailed as to what I did so I won't reiterate that, but the diff is [here](https://github.com/Soostone/katip/compare/49-queue-subscriber?expand=1). One point to mention is that I dropped unagi-chan because my benchmarks for our use cases I found the standard TBQueue from stm to be faster, and given the fact it is in such heavy use and is fairly stable, I am happy to settle on using it.

I think the next step is to dogfood this. I have an app that hasn't quite launched yet that's using katip, so I'll be sure to use this branch going into production with it and see how it does. I don't anticipate any issues and this branch is basically good to go minus maybe some documentation changes.

**Comment by adinapoli** [comment]
> One point to mention is that I dropped unagi-chan because my benchmarks for our use cases I found the standard TBQueue from stm to be faster

Hear, hear! 😀  :trollface: 

> I think the next step is to dogfood this. I have an app that hasn't quite launched yet that's using katip, so I'll be sure to use this branch going into production with it and see how it does. I don't anticipate any issues and this branch is basically good to go minus maybe some documentation changes.

Looking forward to it! Some (hopefully useful) general comments I will dump here about that diff (apparently I cannot attach comments directly, but it could simply be my brain is still half asleep):

* [Here](https://github.com/Soostone/katip/compare/49-queue-subscriber?expand=1#diff-aa7d457cd7e3e425314758cbc9d9aa44R619) are you sure about the comment about safe-exceptions not catching async exceptions? I haven't use the package at all, but [from the README](https://github.com/fpco/safe-exceptions#why-catch-asynchronous-exceptions) it seems that you cannot _recover_ from those, but catching them is recommended for resource cleanup. Maybe it's simply the comment wording that needs to be changed?

* [Here](https://github.com/Soostone/katip/compare/49-queue-subscriber?expand=1#diff-aa7d457cd7e3e425314758cbc9d9aa44R784) when you said that we should write an atomic `tryTBQueue` isn't exactly what the STM monad ensure us? (basically, that if there is something going on with the queue the whole transaction will retry). Maybe you did mean that?

Apart from those minor nitpicks, looks good! (Also, I hope the feedback is useful nevertheless -- it always appreciate people leaving comments on PR of my OSS because it make me feel more connected & involved into the project. Please shout if it doesn't work for you 😀  )

**Comment by MichaelXavier** [comment]
Thanks for the comments, Alfredo!

Haha, the comment on tryWriteTBQueue is just stale, I actually implement that a few lines below. TBQueue only gives us writeTBQueue which tries to write and retries when its full until a slot frees up. This would hang the program in the case of a full queue that isn't being drained. It was simple to add a transactional tryWriteTBQueue that returned immediately on a full queue.

safe-exceptions is capable of catching async exceptions but it makes you opt in by using functions like `tryAsync`, `catchAsync`, etc, whereas vanilla exceptions would catch them by default. You're spot on about cleanup though, async exceptions should always be re-thrown or it breaks things like timeouts, signal handling, OOM, etc. Seems like the main center of cleanup in katip is around closing scribes and calling their finalizers. The user can either call `closeScribe` directly to close a specific scribe, or `closeScribes` more commonly to close all of them. If the user is calling `closeScribe` directly, I think we want to let them deal with any possible exceptions, otherwise we're swallowing exceptions for them, which seems wrong. Because we're calling a cleanup function, if it throws an async or sync exception, there's not really anything else we can do in response. But for `closeScribes` I think we want to probably glue all of these `closeScribe` calls together with `finally`s so that each finalizer gets a chance to run and then an exception can be thrown afterwards if there was one. As the docs say, `finally`, `bracket`, etc actually work on both async and sync exceptions and re-throw both. Semantically, they are dedicated to cleanup, so it is morally fine to use them in the face of any exception, async or sync.

**Comment by adinapoli** [comment]
> Haha, the comment on tryWriteTBQueue is just stale, I actually implement that a few lines below. 

Yea, I saw that, and I was even more confused 😅 

> If the user is calling closeScribe directly, I think we want to let them deal with any possible exceptions, otherwise we're swallowing exceptions for them, which seems wrong.

Absolutely, I agree!

Yep, I think we are on the same page, thanks for clarifying, this looks very good!

**Comment by MichaelXavier** [comment]
Resurrecting this from the dead: I've actually been running this in production for some time successfully and just kind of forgot about it. This is now merged into master, but I'm going to use this time to make a few breaking changes, namely #51 and hopefully #47 so we don't have to thrash the API too much.
